presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -586,4 +592,163 @@ public void prepareFinal()
+            rawInputs.add(new GroupIdPage(squashedGroupIds, page));
+        }
+
+        @Override
+        public void evaluateIntermediate(int groupId, BlockBuilder output)
+        {
+            if (blockBuilders == null) {
+                if (rawInputs == null) {
+                    return;
+                }
+                blockBuilders = new ObjectBigArray<>();  // we need to spill now; let's build the hash table here
+                for (GroupIdPage groupIdPage : rawInputs) {
+                    Page page = groupIdPage.getPage();
+                    GroupByIdBlock groupIdsBlock = groupIdPage.getGroupByIdBlock();
+                    for (int position = 0; position < page.getPositionCount(); position++) {
+                        long currentGroupId = groupIdsBlock.getGroupId(position);
+                        blockBuilders.ensureCapacity(currentGroupId);
+                        RowBlockBuilder rowBlockBuilder = blockBuilders.get(currentGroupId);
+                        if (rowBlockBuilder == null) {
+                            rowBlockBuilder = new RowBlockBuilder(spillingTypes, null, (int) groupIdsBlock.getGroupCount());
+                        }
+
+                        BlockBuilder currentOutput = rowBlockBuilder.beginBlockEntry();
+                        for (int channel = 0; channel < spillingTypes.size(); channel++) {
+                            spillingTypes.get(channel).appendTo(page.getBlock(channel), position, currentOutput);
+                        }
+                        rowBlockBuilder.closeEntry();
+
+                        blockBuilders.set(currentGroupId, rowBlockBuilder);
+                    }
+                }
+                rawInputs = null;
+                rawInputsSize = 0;
+            }
+
+            BlockBuilder singleArrayBlockWriter = output.beginBlockEntry();
+            checkState(rawInputs == null && blockBuilders != null);
+            checkState(singleArrayBlockWriter instanceof SingleArrayBlockWriter);
+
+            RowBlock rowBlock = (RowBlock) blockBuilders.get((long) groupId).build();
+            for (int i = 0; i < rowBlock.getPositionCount(); i++) {
+                singleArrayBlockWriter.appendStructure(rowBlock.getBlock(i));
+            }
+            output.closeEntry();
+        }
+
+        @Override
+        public void evaluateFinal(int groupId, BlockBuilder output)
+        {
+            // checkState(rawInputs == null && blockBuilders == null);
+            delegate.evaluateFinal(groupId, output);
+        }
+
+        @Override
+        public void prepareFinal()
+        {
+            checkState(rawInputs != null && blockBuilders == null);
+            for (GroupIdPage groupIdPage : rawInputs) {
+                delegate.addInput(groupIdPage.getGroupByIdBlock(), groupIdPage.getPage());
+            }
+
+            // cannot set raw inputs to null because evaluateIntermediate may be called after prepareFinal()
```

Comment:
evaluateIntermediate can be called after prepareFinal since it is called by a separate revoking thread from my understanding. This means that we cannot set `rawInputs = null`.

Reply:
If we cannot set it to null, we can still deref it by `rawInputs  = new ..`

Reply:
Unfortunately we cant - this is because the outer function that calls this is buildResult() in hash agg builder - https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java#L306. It tries to build a page out of the result returned by evaluateIntermediate. If there's nothing in the array, it will create a malformed page and fail.

Reply:
If that is case, meaning we shouldn't ever allow memory revoking if an aggregation has started to eval final right?

Reply:
The reason this is not that simple is that memory revoking is done at the driver level. I have implemented a check for this but it may not work.

Take a look at `hasBuiltFinalResult` in `InMemoryHashAggregationBuilder` in my newest iteration.